Restrict subnets only to subnets from regular availability zones in ELB auto-discovery procedure

lobziik opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Outpost, wavelength, and local zone don't support NLB or CLB at the moment. 
This pr adds check for availability zone type and includes only subnets from `availability-zone` typed zones for the auto-discovery.

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #442 

**Special notes for your reviewer**:
I'm not sure about PR type, guess that `feature` is the most suitable one.

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
Added check  for availability zone type within subnets auto-discovery procedure for a load-balancers
Auto-discovery will include only subnets from `availability-zone` typed zones.

IMPORTANT: This change requires an IAM permissions update. IAM policy for control-plane nodes
should be extended with `ec2:DescribeAvailabilityZones` permission.
```

JoelSpeed commented:
/ok-to-test

JoelSpeed commented:
/lgtm

Will need an AWS approver to review as well, but the code looks clean and makes sense to me

lobziik commented:
e2e is failing, i need some time to reproduce and investigate it.
/hold

lobziik commented:
Apparently this requires additional iam permissions for describing availability zones. I need to figure out where it does set within this e2es and give an another thought about this api call error handling.

Will appreciate some advices here, for now i'm not entirely sure how to proceed.

mtulio commented on pkg/providers/v1/aws.go:
```diff
@@ -1156,6 +1164,15 @@ func (s *awsSdkEC2) DescribeSubnets(request *ec2.DescribeSubnetsInput) ([]*ec2.S
 
+func (s *awsSdkEC2) DescribeAvailabilityZones(request *ec2.DescribeAvailabilityZonesInput) ([]*ec2.AvailabilityZone, error) {
+	// AZs are not paged
+	response, err := s.ec2.DescribeAvailabilityZones(request)
```
You may also need to update the [IAM Policy document](https://github.com/kubernetes/cloud-provider-aws/blob/master/docs/prerequisites.md#iam-policies) by adding the `ec2:DescribeAvailabilityZones` permission. That IAM Policy should be attached to the control plane IAM Instance Profile/Role.

lobziik replied:
Updated

lobziik commented:
Updated PR, added a check for the response type. Insufficient permissions does not threat as an error anymore, warning will be raised instead: https://github.com/kubernetes/cloud-provider-aws/pull/499/files#diff-97a31d0f193ea0b4e2c538595daf93405351054377458053a5010f6fdb97f4dcR3539 for provide backward compatability and do not break existing clusters after ccm upgrade.

Added missed permission into documentation also.

lobziik commented:
/hold cancel
/cc @kishorj

lobziik commented:
/cc @kishorj @JoelSpeed 

Sry, I misclicked and break review requests.

JoelSpeed commented on pkg/providers/v1/aws.go:
```diff
@@ -1156,6 +1164,13 @@ func (s *awsSdkEC2) DescribeSubnets(request *ec2.DescribeSubnetsInput) ([]*ec2.S
+func (s *awsSdkEC2) DescribeAvailabilityZones(request *ec2.DescribeAvailabilityZonesInput) ([]*ec2.AvailabilityZone, error) {
+	// AZs are not paged
+	response, err := s.ec2.DescribeAvailabilityZones(request)
+	// do not wrap the error for being able to check its type down the road
```
If you use `fmt.Errorf` with a `%w` format string then in later calls `errors.Is` can check this by unwrapping

JoelSpeed commented on pkg/providers/v1/aws.go:
```diff
@@ -3508,6 +3523,40 @@ func (c *Cloud) findSubnets() ([]*ec2.Subnet, error) {
+
+	azs, err := c.ec2.DescribeAvailabilityZones(azRequest)
+	if err != nil {
+		if awsErr, ok := err.(awserr.RequestFailure); ok && awsErr.Code() == "UnauthorizedOperation" {
```
If you wrap the inner error then you could do something like this to achieve the same
```
awsErr := awserr.RequestFailure{}
if errors.As(err, awsErr) && awsErr.Code() ...
```

lobziik replied:
I looked over the file, and the only pattern to check aws errors is through type assertions. I would prefer to keep it consistent and not introduce a new pattern here. 

But i agree, wrap errors with "%w" and use `errors.As` will be better, but i suggest to address this in other PR and make this same across errors which is returning from this `ec2` sdk wrapper.

JoelSpeed replied:
If this keeps it consistent then fair enough

JoelSpeed reviewed (commented):
Changes look ok, couple of nits about errors but otherwise LGTM, lmk what you think about using the error wrapping, this file already imports the standard library errors package so there's no new dependency on this

JoelSpeed commented:
/lgtm

olemarkus commented:
/lgtm

olemarkus commented:
/pull-kops-e2e-cni-amazonvpc

kishorj commented on pkg/providers/v1/aws.go:
```diff
@@ -3508,6 +3523,40 @@ func (c *Cloud) findSubnets() ([]*ec2.Subnet, error) {
+
+	azs, err := c.ec2.DescribeAvailabilityZones(azRequest)
+	if err != nil {
+		if awsErr, ok := err.(awserr.RequestFailure); ok && awsErr.Code() == "UnauthorizedOperation" {
```
Tbh, if we ignore the error we will keep the existing behavior unless the IAM permissions get updated. I favor throwing the error here instead.

We will have to update the IAM permissions. As for EKS, we can start process to update the managed policies. For kops, we can request the permission to be included.

kishorj replied:
Once the policies get updates, we can backport the changes.

lobziik replied:
I updated pr, now permissions error will be propagated back. 

IAM permissions in kops were extended in https://github.com/kubernetes/kops/pull/14650.

@kishorj can you please take a look? :)

lobziik commented:
I attempted to update permission list for kops. @olemarkus can you please help with that, when you will have time? :) https://github.com/kubernetes/kops/pull/14650 - think it will be a prerequisite for having `pull-cloud-provider-aws-e2e` passed.

olemarkus commented:
/retest

JoelSpeed commented:
Thanks for the fixups @lobziik 

/lgtm

kishorj commented:
Thanks
/lgtm

Need updates to the AmazonEKSClusterPolicy for EKS clusters. Hold merge until we initiate the process to update the managed policy.
